### PR TITLE
feat: add/read features by user projection in openlayers

### PR DESCRIPTION
**Pull request by typebrook**
## Description of Changes

Add/display GeoJSON features with EPSG:4326 coordinates, when custom user projection set by user.

Since `ol/proj.transform()` uses defined projections in module, this PR add more inject items for openlayers adapter.

The following example use patched adapter:
- [EPSG:4326] WGS84
- [EPSG:3857] WGS 84 / Pseudo-Mercator
- [EPSG:3826] (Not built-in projection in `ol/proj`)

## Link to Issue
#319 

[EPSG:4326]: https://outdoorsafetylab.github.io/dumbymap/#%7B%22content%22%3A%22%60%60%60map%5Cnid%3A%204326%5Cnuse%3A%20https%3A%2F%2Funpkg.com%2Fmapclay%400.6.5%2Fdist%2Frenderers%2Fopenlayers.mjs%5Cnproj%3A%20EPSG%3A4326%5Cndraw%3A%20true%5Cncenter%3A%20%5B121%2C24%5D%5Cn%5Cn---%5Cn%5Cnconst%20coords%20%3D%20document.querySelector('%23coords')%3B%5Cnconst%20projection%20%3D%20ol.proj.getUserProjection().getCode()%3B%5Cnmap.on('pointermove'%2C%20e%20%3D%3E%20%7B%5Cn%20%20const%20%5Bx%2C%20y%5D%20%3D%20map.getCoordinateFromPixel(e.pixel)%3B%5Cn%20%20coords.innerHTML%20%3D%20%60%5Cn%20%20%20%20%24%7Bprojection%7D%3Cbr%3E%5Cn%20%20%20%20x%20%3D%3E%20%24%7BNumber(x).toFixed(5)%7D%20%2F%20%5Cn%20%20%20%20y%20%3D%3E%20%24%7BNumber(y).toFixed(5)%7D%5Cn%20%20%60%3B%5Cn%7D)%3B%5Cn%5Cnconst%20textArea%20%3D%20document.querySelector('%23features')%3B%5Cn%5Cndraw.on('change'%2C%20(id%2C_)%20%3D%3E%20%7B%5Cn%20%20const%20features%20%3D%20draw.getSnapshot().filter(f%20%3D%3E%20%5Cn%20%20%20%20f.id.startsWith(target.id)%5Cn%20%20)%3B%5Cn%20%20textArea.textContent%20%3D%20features.length%20%2B%20%5C%22%20features%5C%5Cn%5C%5Cn%5C%22%5Cn%20%20%20%20%2B%20JSON.stringify(features%2C%20null%2C%204)%3B%5Cn%7D)%3B%5Cn%60%60%60%5Cn%5Cn-%20%5BSelect%20%5C%22Points%5C%22%5D(%23Here%20%5C%22%3D%3Eselect%5C%22)%20to%20add%20features%5Cn-%20%5BAnchor%20for%20ref%5D(geo%3A24%2C121)%20%5B121%2C24%5D%5Cn%5Cn%5Cn%3Cdiv%20id%3D%5C%22coords%5C%22%20stye%3D%5C%22background-color%3A%20gray%3B%20height%3A%20200px%3B%5C%22%3E%3C%2Fdiv%3E%5Cn%3Ctextarea%20id%3D%5C%22features%5C%22%20style%3D%5C%22width%3A%20100%25%3B%20height%3A%20500px%5C%22%3E%3C%2Ftextarea%3E%22%7D
[EPSG:3857]: https://outdoorsafetylab.github.io/dumbymap/#%7B%22content%22%3A%22%60%60%60map%5Cnid%3A%203857%5Cnuse%3A%20https%3A%2F%2Funpkg.com%2Fmapclay%400.6.5%2Fdist%2Frenderers%2Fopenlayers.mjs%5Cnproj%3A%20EPSG%3A3857%5Cndraw%3A%20true%5Cncenter%3A%20%5B13469658%2C%202753408%5D%5Cn%5Cn---%5Cnwindow.proj%20%3D%20ol.proj%3B%5Cnconst%20coords%20%3D%20document.querySelector('%23coords')%3B%5Cnconst%20projection%20%3D%20ol.proj.getUserProjection().getCode()%3B%5Cnmap.on('pointermove'%2C%20e%20%3D%3E%20%7B%5Cn%20%20const%20%5Bx%2C%20y%5D%20%3D%20map.getCoordinateFromPixel(e.pixel)%3B%5Cn%20%20coords.innerHTML%20%3D%20%60%5Cn%20%20%20%20%24%7Bprojection%7D%3Cbr%3E%5Cn%20%20%20%20x%20%3D%3E%20%24%7BNumber(x).toFixed(5)%7D%20%2F%20%5Cn%20%20%20%20y%20%3D%3E%20%24%7BNumber(y).toFixed(5)%7D%5Cn%20%20%60%3B%5Cn%7D)%3B%5Cn%5Cnconst%20textArea%20%3D%20document.querySelector('%23features')%3B%5Cn%5Cndraw.on('change'%2C%20(id%2C_)%20%3D%3E%20%7B%5Cn%20%20const%20features%20%3D%20draw.getSnapshot().filter(f%20%3D%3E%20%5Cn%20%20%20%20f.id.startsWith(target.id)%5Cn%20%20)%3B%5Cn%20%20textArea.textContent%20%3D%20features.length%20%2B%20%5C%22%20features%5C%5Cn%5C%5Cn%5C%22%5Cn%20%20%20%20%2B%20JSON.stringify(features%2C%20null%2C%204)%3B%5Cn%7D)%3B%5Cn%60%60%60%5Cn%5Cn-%20%5BSelect%20%5C%22Points%5C%22%5D(%23Here%20%5C%22%3D%3Eselect%5C%22)%20to%20add%20features%5Cn-%20%5BAnchor%20for%20ref%5D(geo%3A2753408%2C13469658)%20%5B13469658%2C%202753408%5D%5Cn%5Cn%5Cn%3Cdiv%20id%3D%5C%22coords%5C%22%20stye%3D%5C%22background-color%3A%20gray%3B%20height%3A%20200px%3B%5C%22%3E%3C%2Fdiv%3E%5Cn%3Ctextarea%20id%3D%5C%22features%5C%22%20style%3D%5C%22width%3A%20100%25%3B%20height%3A%20500px%5C%22%3E%3C%2Ftextarea%3E%22%7D
[EPSG:3826]: https://outdoorsafetylab.github.io/dumbymap/#%7B%22content%22%3A%22%60%60%60map%5Cnid%3A%203826%5Cnuse%3A%20https%3A%2F%2Funpkg.com%2Fmapclay%400.6.5%2Fdist%2Frenderers%2Fopenlayers.mjs%5Cnproj%3A%20EPSG%3A3826%5Cndraw%3A%20true%5Cncenter%3A%20%5B250000%2C%202655023%5D%5Cn%5Cn---%5Cn%5Cnwindow.proj%20%3D%20ol.proj%3B%5Cnconst%20coords%20%3D%20document.querySelector('%23coords')%3B%5Cnconst%20projection%20%3D%20ol.proj.getUserProjection().getCode()%3B%5Cnmap.on('pointermove'%2C%20e%20%3D%3E%20%7B%5Cn%20%20const%20%5Bx%2C%20y%5D%20%3D%20map.getCoordinateFromPixel(e.pixel)%3B%5Cn%20%20coords.innerHTML%20%3D%20%60%5Cn%20%20%20%20%24%7Bprojection%7D%3Cbr%3E%5Cn%20%20%20%20x%20%3D%3E%20%24%7BNumber(x).toFixed(5)%7D%20%2F%20%5Cn%20%20%20%20y%20%3D%3E%20%24%7BNumber(y).toFixed(5)%7D%5Cn%20%20%60%3B%5Cn%7D)%3B%5Cn%5Cnconst%20textArea%20%3D%20document.querySelector('%23features')%3B%5Cn%5Cndraw.on('change'%2C%20(id%2C_)%20%3D%3E%20%7B%5Cn%20%20const%20features%20%3D%20draw.getSnapshot().filter(f%20%3D%3E%20%5Cn%20%20%20%20f.id.startsWith(target.id)%5Cn%20%20)%3B%5Cn%20%20textArea.textContent%20%3D%20features.length%20%2B%20%5C%22%20features%5C%5Cn%5C%5Cn%5C%22%5Cn%20%20%20%20%2B%20JSON.stringify(features%2C%20null%2C%204)%3B%5Cn%7D)%3B%5Cn%60%60%60%5Cn%5Cn-%20%5BSelect%20%5C%22Points%5C%22%5D(%23Here%20%5C%22%3D%3Eselect%5C%22)%20to%20add%20features%5Cn-%20%5BAnchor%20for%20ref%5D(geo%3A2655023%2C250000)%20%5B250000%2C%202655023%5D%5Cn%5Cn%5Cn%3Cdiv%20id%3D%5C%22coords%5C%22%20stye%3D%5C%22background-color%3A%20gray%3B%20height%3A%20200px%3B%5C%22%3E%3C%2Fdiv%3E%5Cn%3Ctextarea%20id%3D%5C%22features%5C%22%20style%3D%5C%22width%3A%20100%25%3B%20height%3A%20500px%5C%22%3E%3C%2Ftextarea%3E%22%7D

### CODE REVIEW

**Review by JamesLMilner — COMMENTED** [review]
@typebrook thanks for raising this - I actually ran into some issues running this locally in the development folder. Would you mind trying out your experiments in there? Check the `development/index.ts` file for details.

I've added some comments on your PR which helped me getting working for me locally with EPSG:4326 and EPSG:3857 working, but wasn't able to get it working with ESPG:3826, but perhaps that  as you said that wasn't expected to work?

File: src/adapters/openlayers.adapter.ts [hunk 1]
```diff
@@ -44,6 +52,9 @@ export class TerraDrawOpenLayersAdapter extends TerraDrawBaseAdapter {
 		this._lib = config.lib;
 
 		this._geoJSONReader = new this._lib.GeoJSON();
+		this._projection = this._lib.getUserProjection() || "EPSG:4326";
```
Comment (JamesLMilner) [inline src/adapters/openlayers.adapter.ts:55]: I had to change this line to `this._projection = this._map.getView().getProjection().getCode() || "EPSG:3857";` and it worked for me.
`

**Comment by typebrook** [comment]
> I actually ran into some issues running this locally

Apologize didn't read DEVELOPMENT guide well, now demo page in `development/` should be fine.

- 2d2e83a update unit tests for key methods `project()` and `unproject()`
- cf8313e resolve your comment, but it is actually for **user projction**, not for **view projection**
- 8486d8d make demo page works on default user projection `EPSG:3857`
- 7996568 add comment-out code for different projections: `EPSG:4326` which is built-in in openlayers, and [`ESRI:100240`](https://epsg.io/102400) which dynamically imported by epsg.io (It is called London Survey Grid, I choose it since demo page is at London)

If you don't need different projection in demo page, you can safely remove 7996568 from this PR.

File: src/adapters/openlayers.adapter.ts (hunk 1 above)
Comment (typebrook) [reply src/adapters/openlayers.adapter.ts]: cf8313e resolve this. `getUserProjection()` return null if user do not intentionally set projection. And openlayers use `EPSG:3857` in transform methods by default.

Your code actually get **view projection**, which is `EPSG:3857` by default, but not related to APIs terra-draw use. 

`getUserProjection()` gets inner property `userProjection` in module `ol/proj`.

It affects **ALL** API methods[^1] including `getCoordinateFromPixel` and `getPixelFromCoordinate`, which are used by `project()` and `unproject` in openlayers adapter. And that's why user should also inject `toLonLat`, `fromLonLat` for getting correct coordinates with projection other than `EPSG:3857`


[^1]: https://github.com/openlayers/openlayers/blob/v10.1.0/src/ol/proj.js#L559

**Comment by JamesLMilner** [comment]
@typebrook thanks for this - I think I wouldn't want to give the impression that we can actively support additional projections, because Terra Draw only officially supports Web Mercator and globe projections at the moment. I'd also want to avoid adding proj4 as dependency if possible.

In your PR, in the development environment if you pass in `getUserProjection`, this will always return `null` and hence you will always get Web Mercator. Am I missing something?

**Comment by typebrook** [comment]
>I wouldn't want to give the impression that we can actively support additional projections

I agree. But "projection" in this case means "CRS" which is used for projection by openlayers

This patch matters because openlayers changes ALL API methods' by user projection. Since `setGeographic()` (which is equals to `setUserProjection("EPSG:4326")`) is a very common use case in Openlayer's example page, adapter need to get key methods to determine how to project/unproject points. (`fromLonLat` and `toLonLat` in this case)

That is, it doesn't matter additional projection is just for EPSG:4326, or for other CRS. User should always pass `fromLonLat` and `toLonLat` to adapter.

> I'd also want to avoid adding proj4 as dependency if possible.

7996568 is just a showcase, you can safely remove it from this PR.

> if you pass in getUserProjection, this will always return null and hence you will always get Web Mercator.

`getUserProjection()` returns property `userProjection`[^1] in `ol/proj`, it is null by default and can be set by `setUserProjection`

```js
setUserProjection('EPSG:4326')
getUserProjection() // Returns Projection { code_: "EPSG:4326", units_: "degrees", extent_: (4) […], worldExtent_: (4) […], axisOrientation_: "neu", global_: true, canWrapX_: true, getPointResolutionFunc_: undefined, defaultTileGrid_: null, metersPerUnit_: 111319.49079327358 }

[^1]: https://github.com/openlayers/openlayers/blob/1e605bc242758bb89e91edc30a26d85f3256c73f/src/ol/proj.js#L550

**Comment by typebrook** [comment]
> adapter need to get key methods to determine how to project/unproject points. (fromLonLat and toLonLat in this case)

Maybe I was wrong. `fromLonLat` and `toLonLat` only get projection internally if ProjecionLike arg is "string". And `getUserProjection()` always return Projection type value. So user don't need to inject these two methods.

https://github.com/openlayers/openlayers/blob/1e605bc242758bb89e91edc30a26d85f3256c73f/src/ol/proj.js#L172-L176
```js
export function get(projectionLike) {
  return typeof projectionLike === 'string'
    ? getProj(/** @type {string} */ (projectionLike))
    : /** @type {Projection} */ (projectionLike) || null;
}
```

Let me do some test.

**Comment by typebrook** [comment]
@JamesLMilner 
Latest commit 56dfb20 removes `fromLonLat` and `toLonLat` from `InjectableOL`, user only need to pass `getUserProjection` to get latest projection in openlayers.

And now `adapter._lib._projection` is a function which always return valid [`Projection`](https://openlayers.org/en/latest/apidoc/module-ol_proj_Projection-Projection.html). So adapter can project/unproject points correctly even if new user projection is set after adapter is build.

**Comment by JamesLMilner** [comment]
Hi @typebrook, thanks for all your work. So I understand, is there a reason not to get the projection from the view rather than from what is set by `setUserProjection`?

I feel like most people would probably set the projection in the view like so:

```typescript
const map = new Map({
	layers: [
		new TileLayer({
			source: new OSM(),
		}),
	],
	target: this.generateId(Libraries.OpenLayers),
	view: new View({
		projection: 'EPSG:3857',
	}),
	controls: [],
});
```

As an example, see here: https://openlayers.org/en/latest/examples/reprojection-wgs84.html

I guess I am trying to understand why we would want to use `getUserProjection` over the set projection for the view? This way you don't need to inject getUserProjection and it will just take the set projection from the view.

**Comment by typebrook** [comment]
First thing first. This issue is not related to **VIEW PROJECTION**. Instead, **USER PROJECTION** affects API methods ues by adapter.

> I guess I am trying to understand why we would want to use getUserProjection

1. adapter use `project/unproject` to process terra-draw GEOJSON feature, which use `ol/Map.getPixelFromCoordinate` and `ol/Map.getCoordinateFromPixel` to transform points from/to current user projection[^1].
2. So here adpater use `ol/Map.fromLonLat` and `ol/Map.toLonLat` to transform points between user projection and WGS84 Lon/Lat. And they need second argument as user projection to do transform [^2], or use `EPSG:3857` by default.
3. `ol/proj` save user projection inside module, so inject getUserProjection is the only way to get its value

I think `user projection` is really a misleading word here... I am not a GIS expert, but I think when we talk about `EPSG:XXXX projection` in this case, it actually means **API now handles data in EPSG:XXXX CRS with current view projection(no matter it is web mercator or global or WGS84), it'll correctly show data in map view**

[^1]: https://github.com/openlayers/openlayers/blob/v10.1.0/src/ol/Map.js#L983
[^2]: https://github.com/openlayers/openlayers/blob/v10.1.0/src/ol/proj.js#L395

**Comment by JamesLMilner** [comment]
@typebrook thanks for clarifying here - I agree the naming from OpenLayers is somewhat confusing. Considering this will only affect people with advanced use cases where they are not using Web Mercator projection, I think this is probably okay to merge. I appreciate your persistence here, apologies if it took a while to understand what you were saying.

**Comment by typebrook** [comment]
@JamesLMilner Thanks for your detailed review. Terra-draw save me a lot of time on my ongoing project. It is awesome.

> apologies if it took a while to understand what you were saying

No worry, It also took me time to realize how it works. I should mention at the first time that: user-projection equals to the CRS API is working with.

**Pull request merged**